testcase/stale-read/client.go

```diff
@@ -0,0 +1,306 @@
+	log.Info("[stale read] start to test...")
+	var wg sync.WaitGroup
+	// Write data
+	for i := 0; i < s.Concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			rnd := rand.New(rand.NewSource(time.Now().Unix()))
+			for {
+				ids := make([]int, rand.Intn(99)+1)
+				for i := range ids {
+					ids[i] = rand.Intn(s.TotalRows) + 1
+				}
+				idsStr := strings.Trim(strings.Join(strings.Fields(fmt.Sprint(ids)), ", "), "[]")
+				pad := make([]byte, 255)
+				util.RandString(pad, rnd)
+				stmt := fmt.Sprintf(`UPDATE %s.%s SET pad="%s" WHERE id in (%s)`, s.DBName, Table, pad, idsStr)
+				if _, err := s.db.Exec(stmt); err != nil {
+					log.Warnf("[stale read] Failed to execute sql [%s], error: %v", stmt, err)
+				}
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					time.Sleep(s.RequestInterval)
+				}
+			}
+		}()
+	}
+	// Wait `MaxStaleness` elapsed so all read can reture data
+	time.Sleep(time.Duration(s.MaxStaleness) * time.Second)
+	// Read data
+	for i := 0; i < s.Concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			s.mustEqualRead(ctx)
+		}()
+	}
+	wg.Wait()
+	return nil
+}
+
+func (s *staleRead) mustEqualRead(ctx context.Context) {
+	conn1, err := s.db.Conn(ctx)
+	if err != nil {
+		return
+	}
+	conn2, err := s.db.Conn(ctx)
+	if err != nil {
+		return
+	}
+	count := 0
+	for {
+		// prepare id
+		ids := make([]int, rand.Intn(99)+1)
+		for i := range ids {
+			ids[i] = rand.Intn(s.TotalRows) + 1
+		}
+		// 0s-100s ago
+		ago := rand.Intn(s.MaxStaleness)
+		timeAgo := time.Now().Add(time.Duration(-ago) * time.Second)
+		stale, ts, err1 := s.staleRead(ctx, conn1, timeAgo, ids)
+		strong, err2 := s.strongRead(ctx, conn2, ts, ids)
+		if err1 == nil && err2 == nil {
+			// The result should be the same
+			if len(stale) != len(strong) || len(strong) == 0 {
+				log.Fatalf("got different number of resluts at %d ago, ids: %d, stale: %d, strong: %d, stale res: %v, strong res: %v", ago, len(ids), len(stale), len(strong), stale, strong)
+			}
+			for id, pad := range strong {
+				if stalePad, ok := stale[id]; !ok || stalePad != pad {
+					log.Fatalf("got different resluts for id %d at %ds ago, tso: %v, stale read: %s, strong read: %s", id, ago, ts, stale[id], pad)
+				}
+			}
+			if count%10000 == 0 {
+				sql := fmt.Sprintf("select * from %s.%s where id in (%s)", s.DBName, Table, strings.Trim(strings.Join(strings.Fields(fmt.Sprint(ids)), ", "), "[]"))
+				log.Infof("Stale read %d seconds ago equal to strong read, sql: %s", ago, sql)
+			}
+		} else {
+			log.Warnf("read return error, stale: %v, strong: %v", err1, err2)
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			time.Sleep(s.RequestInterval)
+			count++
+		}
+	}
+}
+
+func (s *staleRead) staleRead(ctx context.Context, conn *sql.Conn, t time.Time, ids []int) (map[int]string, uint64, error) {
+	txn, err := conn.BeginTx(ctx, &sql.TxOptions{})
```

Comment:
`BeginTx` here will directly send `begin` to sql server. I think we should use `ExecuteContext` to begin a stale transaction by `conn.BeginTx(ctx, "START TRANSACTION READ ONLY AS OF TIMESTAMP...") `

Reply:
`conn.BeginTx(ctx, "START TRANSACTION READ ONLY AS OF TIMESTAMP...")` not match the function signature `func (*sql.Conn).BeginTx(ctx context.Context, opts *sql.TxOptions)`, I use the `SET TRANSACTION READ ONLY AS OF TIMESTAMP` instead, PTAL